File: ios/LimePrintBridge/Bridge/QzMessageRouter.swift
import Foundation

/// Dispatches JSON envelopes from qz-tray.js to the appropriate handler.
///
/// Mirrors the call dispatch in `src/qz/ws/PrintSocketClient.processMessage()`
/// and the SocketMethod enum in `src/qz/ws/SocketMethod.java` (Java reference
/// in this same repository).
///
/// Transport-agnostic: the per-connection `send` closure is supplied by
/// `QzBridge` (which pushes back into the WKWebView via
/// `evaluateJavaScript("__qzBridge.recv(...)")`). MVP signature handling: we
/// accept the certificate on first connect (with user dialog) and skip
/// per-call signature verification.
@MainActor
final class QzMessageRouter {

    typealias ApprovalHandler = (_ pem: String, _ commonName: String?) async -> Bool

    private let printer: EscPosPrinter
    private let registry: PrinterRegistry
    private let allowedCerts: QzAllowedCerts
    private let approvalHandler: ApprovalHandler

    private struct ConnState { var certPem: String? = nil; var allowed: Bool = false }
    private var connections: [String: ConnState] = [:]

    private static let emulatedVersion = "2.2.6"

    init(printer: EscPosPrinter,
         registry: PrinterRegistry,
         allowedCerts: QzAllowedCerts,
         approvalHandler: @escaping ApprovalHandler) {
        self.printer = printer
        self.registry = registry
        self.allowedCerts = allowedCerts
        self.approvalHandler = approvalHandler
    }

    func onConnectionOpened(id: String) {
        connections[id] = ConnState()
    }

    func onConnectionClosed(id: String) {
        connections.removeValue(forKey: id)
    }

    /// Plain-text probe handshake (qz-tray.js sends "getProgramName" before any JSON).
    /// Mirrors `Constants.PROBE_REQUEST` / `PROBE_RESPONSE` in
    /// `src/qz/common/Constants.java:63-64`.
    func handlePlainText(_ message: String, on id: String, send: @escaping (String) -> Void) -> Bool {
        switch message {
        case "getProgramName":
            send("QZ Tray"); return true
        case "ping":
            return true
        default:
            return false
        }
    }

    func dispatch(_ json: [String: Any], on id: String, send: @escaping (String) -> Void) async {
        let uid = json["uid"] as? String ?? ""

        if let pem = json["certificate"] as? String, !pem.isEmpty {
            await handleCertHandshake(uid: uid, pem: pem, on: id, send: send)
            return
        }

        guard let call = json["call"] as? String, !call.isEmpty else {
            sendError(uid: uid, message: "Missing 'call' in message", send: send)
            return
        }

        // Unsigned calls don't require cert handshake
        switch call {
        case "getVersion":
            sendResult(uid: uid, result: Self.emulatedVersion, send: send); return
        case "websocket.getNetworkInfo", "networking.deviceLegacy", "networking.device":
            sendResult(uid: uid, result: ["ipAddress": "127.0.0.1", "macAddress": "00:00:00:00:00:00"], send: send); return
        case "websocket.getNetworkInterfaces", "networking.devices":
            sendResult(uid: uid, result: [Any](), send: send); return
        default: break
        }

        guard let state = connections[id], state.allowed else {
            sendError(uid: uid, message: "Connection not yet authorized", send: send)
            return
        }

        do {
            try await handleAuthorizedCall(call: call, uid: uid, json: json, send: send)
        } catch {
            sendError(uid: uid, message: error.localizedDescription, send: send)
        }
    }

    private func handleAuthorizedCall(call: String,
                                      uid: String,
                                      json: [String: Any],
                                      send: @escaping (String) -> Void) async throws {
        switch call {
        case "printers.getDefault":
            let name = registry.saved()?.name as Any? ?? NSNull()
            sendResult(uid: uid, result: name, send: send)

        case "printers.find":
            let query = (json["params"] as? [String: Any])?["query"] as? String
            let names: [String] = registry.saved().map { [$0.name] } ?? []
            let filtered = (query?.isEmpty ?? true)
                ? names
                : names.filter { $0.localizedCaseInsensitiveContains(query ?? "") }
            sendResult(uid: uid, result: filtered, send: send)

        case "printers.details", "printers.detail":
            if let s = registry.saved() {
                sendResult(uid: uid, result: [["name": s.name, "driver": "ESCPOS", "connection": "BLE"]], send: send)
            } else {
                sendResult(uid: uid, result: [Any](), send: send)
            }

        case "printers.getStatus":
            sendResult(uid: uid, result: NSNull(), send: send)

        case "print":
            guard let params = json["params"] as? [String: Any] else {
                throw NSError(domain: "QzMessageRouter", code: 1,
                              userInfo: [NSLocalizedDescriptionKey: "Missing print params"])
            }
            try await printer.print(params: params)
            sendResult(uid: uid, result: NSNull(), send: send)

        case "websocket.stop":
            sendResult(uid: uid, result: NSNull(), send: send)

        default:
            NSLog("QzMessageRouter: unhandled call=\(call)")
            sendResult(uid: uid, result: NSNull(), send: send)
        }
    }

    private func handleCertHandshake(uid: String, pem: String, on id: String, send: @escaping (String) -> Void) async {
        var state = connections[id] ?? ConnState()
        state.certPem = pem
        connections[id] = state

        switch allowedCerts.decision(for: pem) {
        case .allow:
            connections[id]?.allowed = true
            sendResult(uid: uid, result: NSNull(), send: send)
        case .block:
            sendError(uid: uid, message: "Connection blocked by user", send: send)
        case .unknown:
            let cn = extractCommonName(from: pem)
            let approved = await approvalHandler(pem, cn)
            if approved {
                connections[id]?.allowed = true
                sendResult(uid: uid, result: NSNull(), send: send)
            } else {
                sendError(uid: uid, message: "Connection blocked by user", send: send)
            }
        }
    }

    // MARK: - Encoding helpers

    private func sendResult(uid: String, result: Any, send: (String) -> Void) {
        let env: [String: Any] = ["uid": uid, "result": result]
        if let data = try? JSONSerialization.data(withJSONObject: env, options: [.fragmentsAllowed]),
           let s = String(data: data, encoding: .utf8) {
            send(s)
        }
    }

    private func sendError(uid: String, message: String, send: (String) -> Void) {
        let env: [String: Any] = ["uid": uid, "error": message]
        if let data = try? JSONSerialization.data(withJSONObject: env, options: [.fragmentsAllowed]),
           let s = String(data: data, encoding: .utf8) {
            send(s)
        }
    }

    private func extractCommonName(from pem: String) -> String? {
        let cleaned = pem
            .replacingOccurrences(of: "-----BEGIN CERTIFICATE-----", with: "")
            .replacingOccurrences(of: "-----END CERTIFICATE-----", with: "")
            .components(separatedBy: .whitespacesAndNewlines).joined()
        guard
            let der = Data(base64Encoded: cleaned),
            let cert = SecCertificateCreateWithData(nil, der as CFData)
        else { return nil }
        return SecCertificateCopySubjectSummary(cert) as String?
    }
}

File: android/app/src/main/java/com/limebooking/printbridge/qz/QzMessageRouter.kt
package com.limebooking.printbridge.qz

import android.util.Log
import com.limebooking.printbridge.printer.EscPosPrinter
import com.limebooking.printbridge.printer.PrinterRegistry
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors

/**
 * Dispatches JSON envelopes from qz-tray.js to the appropriate handler.
 *
 * Mirrors the call dispatch in src/qz/ws/PrintSocketClient.processMessage()
 * and the SocketMethod enum in src/qz/ws/SocketMethod.java.
 *
 * Transport-agnostic: works with both [QzServer] (real WebSocket) and the
 * in-WebView [QzBridge] (JavascriptInterface).
 *
 * MVP signature handling: we accept the certificate on first connect (with
 * user dialog) and skip per-call signature verification. Localhost-only
 * attack surface — only our own WebView can reach this dispatcher.
 */
class QzMessageRouter(
    private val printerRegistry: PrinterRegistry,
    private val printer: EscPosPrinter,
    private val allowedCerts: QzAllowedCerts,
    private val certApprovalHandler: CertApprovalHandler
) {

    companion object {
        private const val TAG = "QzMessageRouter"
        private const val EMULATED_VERSION = "2.2.6"
    }

    /** Asks the user to approve an unknown PWA cert. */
    fun interface CertApprovalHandler {
        /** Blocks until the user decides; returns true if allowed. */
        fun requestApproval(certPem: String, commonName: String?): Boolean
    }

    private data class ConnState(var certPem: String? = null, var allowed: Boolean = false)

    private val connections = ConcurrentHashMap<String, ConnState>()
    private val pool = Executors.newCachedThreadPool { r ->
        Thread(r, "qz-worker").apply { isDaemon = true }
    }

    fun onConnectionOpened(conn: QzConnection) {
        connections[conn.id] = ConnState()
    }

    fun onConnectionClosed(conn: QzConnection) {
        connections.remove(conn.id)
    }

    fun dispatch(conn: QzConnection, json: JSONObject) {
        val uid = json.optString("uid", "")

        if (json.has("certificate")) {
            handleCertHandshake(conn, uid, json.optString("certificate"))
            return
        }

        val call = json.optString("call")
        if (call.isEmpty()) {
            sendError(conn, uid, "Missing 'call' in message")
            return
        }

        pool.execute {
            try {
                handleCall(conn, uid, call, json)
            } catch (e: Exception) {
                Log.e(TAG, "Error processing call=$call", e)
                sendError(conn, uid, e.message ?: e.javaClass.simpleName)
            }
        }
    }

    private fun handleCertHandshake(conn: QzConnection, uid: String, pem: String) {
        if (pem.isEmpty()) {
            connections[conn.id]?.allowed = true
            sendResult(conn, uid, JSONObject.NULL)
            return
        }

        val state = connections[conn.id] ?: ConnState().also { connections[conn.id] = it }
        state.certPem = pem

        when (allowedCerts.decisionFor(pem)) {
            QzAllowedCerts.Decision.ALLOW -> {
                state.allowed = true
                sendResult(conn, uid, JSONObject.NULL)
            }
            QzAllowedCerts.Decision.BLOCK -> {
                sendError(conn, uid, "Connection blocked by user")
                conn.close()
            }
            QzAllowedCerts.Decision.UNKNOWN -> pool.execute {
                val cn = extractCommonName(pem)
                val approved = try {
                    certApprovalHandler.requestApproval(pem, cn)
                } catch (e: Exception) {
                    Log.w(TAG, "Approval handler failed", e); false
                }
                if (approved) {
                    state.allowed = true
                    sendResult(conn, uid, JSONObject.NULL)
                } else {
                    sendError(conn, uid, "Connection blocked by user")
                    conn.close()
                }
            }
        }
    }

    private fun handleCall(conn: QzConnection, uid: String, call: String, json: JSONObject) {
        // Unsigned calls don't require cert handshake
        when (call) {
            "getVersion" -> {
                sendResult(conn, uid, EMULATED_VERSION); return
            }
            "websocket.getNetworkInfo", "networking.deviceLegacy", "networking.device" -> {
                sendResult(
                    conn, uid,
                    JSONObject().put("ipAddress", "127.0.0.1").put("macAddress", "00:00:00:00:00:00")
                ); return
            }
            "websocket.getNetworkInterfaces", "networking.devices" -> {
                sendResult(conn, uid, JSONArray()); return
            }
        }

        val state = connections[conn.id]
        if (state?.allowed != true) {
            sendError(conn, uid, "Connection not yet authorized")
            return
        }

        when (call) {
            "printers.getDefault" -> {
                sendResult(conn, uid, printerRegistry.defaultPrinterName() ?: JSONObject.NULL)
            }
            "printers.find" -> {
                val query = json.optJSONObject("params")?.optString("query")
                val matches = printerRegistry.listAvailable().map { it.name }.filter {
                    query.isNullOrEmpty() || it.contains(query, ignoreCase = true)
                }
                sendResult(conn, uid, JSONArray(matches))
            }
            "printers.details", "printers.detail" -> {
                val arr = JSONArray()
                printerRegistry.listAvailable().forEach { p ->
                    arr.put(JSONObject().put("name", p.name).put("driver", "ESCPOS").put("connection", "BLUETOOTH"))
                }
                sendResult(conn, uid, arr)
            }
            "printers.getStatus" -> sendResult(conn, uid, JSONObject.NULL)
            "print" -> {
                val params = json.getJSONObject("params")
                printer.print(params)
                sendResult(conn, uid, JSONObject.NULL)
            }
            "websocket.stop" -> sendResult(conn, uid, JSONObject.NULL)
            else -> {
                Log.w(TAG, "Unhandled call: $call (responding null)")
                sendResult(conn, uid, JSONObject.NULL)
            }
        }
    }

    private fun sendResult(conn: QzConnection, uid: String, result: Any) {
        try {
            conn.send(JSONObject().put("uid", uid).put("result", result).toString())
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send result", e)
        }
    }

    private fun sendError(conn: QzConnection, uid: String?, message: String) {
        try {
            conn.send(JSONObject().put("uid", uid ?: "").put("error", message).toString())
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send error", e)
        }
    }

    private fun extractCommonName(pem: String): String? {
        return try {
            val cleaned = pem.replace("-----BEGIN CERTIFICATE-----", "")
                .replace("-----END CERTIFICATE-----", "")
                .replace("\\s".toRegex(), "")
            val der = android.util.Base64.decode(cleaned, android.util.Base64.DEFAULT)
            val factory = java.security.cert.CertificateFactory.getInstance("X.509")
            val cert = factory.generateCertificate(java.io.ByteArrayInputStream(der))
                    as java.security.cert.X509Certificate
            val dn = cert.subjectX500Principal.name
            Regex("CN=([^,]+)").find(dn)?.groupValues?.get(1)
        } catch (e: Exception) {
            Log.w(TAG, "Failed to extract CN from cert", e); null
        }
    }
}
